network/modules/tcp/server_socket.py

```diff
@@ -0,0 +1,114 @@
+            Could also use socket.gethostname(). (needed to enable other machines to connect)
+        port: int (default 5000)
+            The host combined with the port will form an address (e.g. localhost:5000)
+        connection_timeout: float (default 10.0)
+            Timeout for operations such as recieve
+
+        Returns
+        -------
+        tuple[bool, TcpServerSocket | None]
+            The first parameter represents if the socket creation is successful.
+            - If it is not successful, the second parameter will be None.
+            - If it is successful, the second parameter will be the created
+              TcpServerSocket object.
+        """
+
+        # Reassign instance before check or Pylance will complain
+        socket_instance = instance
+        if socket_instance is not None:
+            return True, TcpServerSocket(cls.__create_key, socket_instance)
+
+        if socket.has_dualstack_ipv6():
+            # Create server which can accept both IPv6 and IPv4 if possible
+            try:
+                server = socket.create_server(
+                    (host, port),
+                    family=socket.AF_INET6,
+                    dualstack_ipv6=True,
+                )
+            except socket.gaierror as e:
+                print(
+                    f"Could not connect to socket, address related error: {e}.",
+                    "Make sure the host and port are correct.",
+                )
+                return False, None
+            except socket.error as e:
+                print(f"Could not connect to socket, connection error: {e}.")
+                return False, None
+        else:
+            # Otherwise, server can only accept IPv4
+            try:
+                server = socket.create_server((host, port))
+            except socket.gaierror as e:
+                print(
+                    f"Could not connect to socket, address related error: {e}. "
+                    "Make sure the host and port are correct."
+                )
+                return False, None
+            except socket.error as e:
+                print(f"Could not connect to socket, connection error: {e}.")
+                return False, None
+
+        # Currently listening, waiting for a connection
+        if host == "":
+            print(f"Listening for external connections on port {port}")
+        else:
+            print(f"Listening for internal connections on {host}:{port}")
+
+        server.settimeout(connection_timeout)
+        # This is in blocking mode, nothing can happen until this finishes, even keyboard interrupt
+        socket_instance, addr = server.accept()
+
+        # Now, a connection as been accepted (created a new 'client' socket)
+        print(f"Accepted a connection from {addr[0]}:{addr[1]}")
+
+        socket_instance.settimeout(connection_timeout)
+
+        server.close()
+        print("No longer accepting new connections.")
```

Comment:
why do we connect a client, close the server, and then return the client socket? isn't this class meant to return a server socket?

Reply:
It might look a little confusing, but TCP is a 1-to-1 connection protocol, so there is no difference between the two communicating sockets (here they are the client sockets). The "server" socket just listens for connection requests, and then sets up the connection. It does not participate in the communication.